File: src/pokemonStorageSystem/PokemonStorageSystem.java
package pokemonStorageSystem;

import basicFunctions.UserInput;
import java.util.Scanner;
import pokemon.Pokemon;
import ui.GameManager;

/**
 *
 * @author ellen
 */
public class PokemonStorageSystem {
    private static Scanner keyInput;
    private static final String TITLE = "Pokemon Storage System:";
    private static final String[] CHOICES = {"Deposit Pokémon",
        "Pick Up Pokémon",
        "Switch Pokémon",
        "Move Items",
        "Back"};
    
    private static PokemonStorageSystem pokemonStorageSystem;
    private static Pokemon[][] storage;
    private static String[] boxNames;
    
    private PokemonStorageSystem() {
        storage = new Pokemon[31][30];
        boxNames = new String[31];
        setInitialBoxNames();
        keyInput = new Scanner(System.in);
    }
    
    /**
     * Returns the singleton instance of PokemonStorageSystem. If the instance 
     * doesn't exist, it will be created.
     * 
     * @return singleton instance
     */
    public static PokemonStorageSystem getInstance() {
        if (pokemonStorageSystem == null) {
            pokemonStorageSystem = new PokemonStorageSystem();
        }
        return pokemonStorageSystem;
    }
    
    private void setInitialBoxNames() {
        for (int i = 0; i < boxNames.length; i++) {
            boxNames[i] = "BOX " + (i + 1);
        }
    }
    
    /**
     * Set the name of a box in the storage system.
     * @param boxIndex the index of the box for which the name will be changed.
     * @param boxName the new name to set
     */
    public void setBoxName(int boxIndex, String boxName) {
        boxNames[boxIndex] = boxName;
    }
    
    public void accessPokemonStorageSystem() {
        int choice = UserInput.makeMenu(keyInput, TITLE, CHOICES) - 1;
        while (choice != CHOICES.length - 1) {
            switch(choice) {
                case 0:
                    depositPokemon();
                    break;
                case 1:
                    getPokemon();
                    break;
                case 2:
                    switchPokemon();
                    break;
                case 3:
                    moveItems();
                    break;
            }
        }
    }
    
    /**
     * Deposit Pokemon from the party into the storage boxes.
     */
    public void depositPokemon() {
        int pkmnIndex = UserInput.makeMenu(keyInput, "Which Pokémon?", 
                UserInput.convertToMakeMenuArray(GameManager.getPartyPokemonList()));
        Pokemon p = GameManager.getPlayer().getPartyMember(pkmnIndex);
        depositPokemon(p);
    }
    
    private void depositPokemon(Pokemon p) {
        int boxIndex = viewBoxList();
        if (addToBox(boxIndex, p)) {
            UserInput.printALine(keyInput, p.getName() + " was placed in " + 
                    boxNames[boxIndex] + ".");
        } else {
            int choice = UserInput.makeMenu(keyInput,
                    boxNames[boxIndex] + " is full. Choose another box?", 
                    UserInput.CHOICES);
            if (choice == UserInput.YES) {
                depositPokemon(p);
            }
        }
    }
    
    /**
     * Get Pokemon from storage into the party. If there is no space in the
     * party, the action cannot be performed.
     * 
     * TODO name v nickname
     */
    public void getPokemon() {
        int boxIndex = viewBoxList();
        int pkmnIndex = viewBoxContents(boxIndex);
        if (GameManager.getPlayer().addToParty(storage[boxIndex][pkmnIndex])) {
            UserInput.printALine(keyInput, storage[boxIndex][pkmnIndex].getName() + 
                    " was added to " + GameManager.getPlayer().getName() + "'s party.");
        }
    }
    
    /**
     * Move Pokemon freely between the party and the storage system.
     */
    public void switchPokemon() {
        
    }
    
    /**
     * View the items held by Pokemon in storage and in the party, and move
     * these freely. No Pokemon can be moved.
     */
    public void moveItems() {
        
    }
    
    /**
     * Print a menu of names of all boxes in the storage system.
     * @return the index of the selected box
     */
    private int viewBoxList() {
        return UserInput.makeMenu(keyInput, TITLE, 
                UserInput.convertToMakeMenuArray(boxNames));
    }
    
    /**
     * Print a menu of all Pokémon within the selected box.
     * @param boxIndex the index of the box to display
     * @return the index of the selected Pokémon
     */
    private int viewBoxContents(int boxIndex) {
        String[] pkmn = new String[30];
        for (int i = 0; i < pkmn.length; i++) {
            pkmn[i] = storage[boxIndex][i].toString();
        }
        return UserInput.makeMenu(keyInput, TITLE,
                UserInput.convertToMakeMenuArray(pkmn));
    }
    
    /**
     * Checks if there is room in the selected box, and adds the Pokémon to
     * the first available spot.
     * @param boxIndex index of the box to add to
     * @param p the Pokémon to add to the box
     * @return true if added successfully, false if not
     */
    private boolean addToBox(int boxIndex, Pokemon p) {
        for (int i = 0; i < storage[boxIndex].length; i++) {
            if (storage[boxIndex][i] == null) {
                storage[boxIndex][i] = p;
                return true;
            }
        }
        return false;
    }
}

File: pysrc/pokemonstoragesystem.py
import userinput

TITLE = "Pokemon Storage System:"
CHOICES = ["Deposit Pokémon",
           "Pick Up Pokémon",
           "Switch Pokémon",
           "Move Items"]

NUM_BOXES = 18

class Box:
    def __init__(self, name):
        self.name = name
        self.pokemon = []

    def set_name(self, name):
        if name in storage:
            raise ValueError("Duplicate name")
        storage[name] = self
        if self.name in storage:
            del storage[self.name]
        self.name = name


storage = {}

def set_initial_boxes():
    for i in range(NUM_BOXES):
        name = "BOX {0}".format(i + 1)
        storage[name] = Box(name)
    

def rename(old_name, new_name):
    ##todo decide where to handle exceptions (keyerror, valueerror)
    storage[old_name].set_name(new_name)


def access_system():
    choice = userinput.make_title_menu(TITLE, CHOICES)
    ##todo this should be an action menu


def deposit_pokemon():
    pkmn_index = userinput.make_title_menu("Which Pokémon?",
                                           gamemanager.get_party_pkmn_list)
    deposit(gamemanager.get_player().get_party_member(pkmn_index))


def deposit(pkmn):
    box_idx = view_box_list()
    if add_to_box(box_idx, pkmn):
        input("{0} was placed in {1}.".format(pkmn.get_name(), 0)) ##TODO
    else:
        choice = userinput.make_menu(
            "{0} is full. Choose another box?".format(1), ##TODO
            userinput.CHOICES)
        if choice == userinput.YES:
            deposit(pkmn)


def get_pokemon():
    box_idx = view_box_list()
    pkmn_idx = view_box_contents(box_idx)
    if gamemanager.get_player().add_to_party(storage[box_idx][pkmn_idx]):
        input("{0} was added to {1}'s party.".format(
            storage[box_idx][pkmn_idx].get_name(),
            gamemanager.get_player().get_name()))


def switch_pkmn():
    pass


def move_items():
    pass


def view_box_list():
    return userinput.make_title_menu(TITLE, storage) ##will these print right?


def view_box_contents():
    pass


def add_to_box(box_idx, pkmn):
    pass
